docs: add docs for projections (#18056)
druid-io/druid: docs/querying/projections.md

  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

Projections are a type of aggregation that is computed and stored as part of a segment. The pre-aggregated data can speed up queries by reducing the number of rows that need to be processed for any query shape that matches a projection. 

## Create a projection

A projection has three components:

- Virtual columns (`spec.projections.virtualColumns`) that are used to compute a projection. The source data for the virtual columns must exist in your datasource.
- Grouping columns (`spec.projections.groupingColumns`) that are used to group a projection. They must either already exist in your datasource or be defined in `virtualColumns`. The order in which you define your grouping columns equates to the order in which data is sorted in the projection, always ascending.

> - Aggregators (`spec.projections.aggregators`) that define the columns you want to create projections for and which aggregator to use for that column. They must either already exist in your datasource or be defined in `virtualColumns`.

Reviewer: ```suggestion
- Aggregators (`spec.projections.aggregators`) that define the columns to create projections for and the aggregator to use for that column. The source columns? must either already exist in your datasource or be defined in `virtualColumns`.
```
" columns you want to create projections for ": is this the "source data" ? we were talking about in line 36? "

The aggregators are what Druid attempts to match when you run a query. If an aggregator in a query matches an aggregator you defined in your projection, Druid uses it.

You can either create a projection at ingestion time or after the datasource is created. 

> Note that any projection dimension you create becomes part of your datasource. To remove a projection from your datasource, you need to reingest the data. Alternatively, you can use a query context parameter to not use projections for a specific query.

Reviewer: ```suggestion
Note that any projection dimension you create becomes part of your datasource. You must you reingest your  data to remove a projection from your datasource. Alternatively, you can use a query context parameter to avoid using projections for a specific query.
```

### At ingestion time

> To create a projection at ingestion time, use the [`projectionsSpec` block in your ingestion spec](../ingestion/ingestion-spec.md#projections).

Reviewer: is this just native json ingestion? I don't think of MSQ ingestion having an "ingestion spec"

I feel like both of these need examples.

Author: yeah, there's an example in the linked page. i can copy it over. an msq example is on my to do list

To create projections for SQL-based ingestion, you need to also have the [`druid-catalog`](../development/extensions-core/catalog.md) extension loaded.

> ### After ingestion

Reviewer: should this be an existing table vs new table

> You can define a projection after you ingest data. Although you can define the projection in a compaction spec, we recommend using the [`druid-catalog`](../development/extensions-core/catalog.md) extension.

Reviewer: Switch this around to mention the recommended way first. Then say if that way doesn't work ,you can optionally use the less-preferred message.

> The following API call includes a payload with the `properties.projections` block that defines your projections:

Reviewer: I like this example.

              "type": "string",
              "name": "channel",
              "multiValueHandling": "SORTED_ARRAY",
              "createBitmapIndex": true
            },
            {
              "type": "string",
              "name": "page",
              "multiValueHandling": "SORTED_ARRAY",
              "createBitmapIndex": true
            }
          ],
          "aggregators": [
            {
              "type": "HLLSketchBuild",
              "name": "distinct_users",
              "fieldName": "user",
              "lgK": 12,
              "tgtHllType": "HLL_4"
            },
            {
              "type": "longSum",
              "name": "sum_added",
              "fieldName": "added"
            },
            {
              "type": "longSum",
              "name": "sum_deleted",
              "fieldName": "deleted"
            }
          ]
        }
      }
    ]
  }
}
```

</details>

In this example, Druid aggregates data into `distinct_user`, `sum_added`, and `sum_deleted` dimensions based on the aggregator that's specified and a source dimension. These aggregations are grouped by the columns you define in `groupingColumns`.

## Use a projection 

> Druid automatically uses a projection if your query matches a projection you've defined. There are some query context parameters that give you some control on how projections are used and Druid's behavior:

Reviewer: ```suggestion
Druid automatically uses a projection if your query matches a projection you've defined. You can use the following query context parameters do override  the default behavior for using projections:
```

> - `useProjection`: The name of a projection you defined. The query engine must use that projection and will fail the query if the projection does not match the query.

Reviewer: ```suggestion
- `useProjection`: The name of a projection. Set to `true`  to require the query engine to use a specific projection. Otherwise the query fails when no matching projection exists. No default.
```

> - `forceProjections` `true` or `false`. The query engine must use a projection and will fail the query if there isn't a matching projection.

Reviewer: ```suggestion
- `forceProjections`: Set to `true`  to require the query engine to use a projection. Otherwise the query fails when no matching projection exists. Defaults to `false`.
```

> -  `noProjections`:  `true` or `false`. The query engine won't use any projections.

Reviewer: ```suggestion
- `noProjections`:  Set to `true` to prevent the query engine from using projections altogether. Defaults to `false`.
```
seriously this should be useProjections which defaults to true and you could set to false to disable. @clintropolis :/

## Compaction

> To use compaction on a datasource that includes projections, you need to set the type to catalog: `spec.type: catalog`:

Reviewer: ```suggestion
To use compaction on a datasource that includes projections, you need to set the spec type to catalog: `spec.type: catalog`:
```
